File: API-Proyecto/Data/Repositories/PedidoRepository.cs
using Domain.Entities;
using Domain.Interfaces.Repositories;

namespace Data.Repositories
{
    /// <summary>
    /// Repositorio de Pedido con datos mock en memoria.
    /// Implementa la interfaz IPedidoRepository.
    /// </summary>
    public class PedidoRepository : IPedidoRepository
    {
        // Lista estática en memoria que simula la tabla PEDIDO
        private static List<Pedido> listaPedidos = new List<Pedido>
        {
            new Pedido(1, 1, 1, new DateTime(2025, 1, 15), "pedido", "Pedido urgente de tornillos", false),
            new Pedido(2, 1, 2, new DateTime(2025, 1, 20), "enviado", "Herramientas para taller", false),
            new Pedido(3, 2, 3, new DateTime(2025, 2, 1), "entregado", "Material de construcción", false),
            new Pedido(4, 2, 4, new DateTime(2025, 2, 5), "pedido", "Suministros de oficina", true),
            new Pedido(5, 3, 5, new DateTime(2025, 2, 8), "enviado", "Repuestos de maquinaria", false)
        };

        // Contador para generar IDs automáticamente
        private static int contadorId = 6;

        /// <summary>
        /// Obtiene la lista completa de pedidos.
        /// </summary>
        /// <returns>Lista de pedidos</returns>
        public List<Pedido> GetListaPedidos()
        {
            return listaPedidos;
        }

        /// <summary>
        /// Obtiene la lista de pedidos realizados por un usuario.
        /// </summary>
        /// <param name="idUsuario">ID del usuario</param>
        /// <returns>Lista de pedidos del usuario</returns>
        public List<Pedido> GetListaPedidosPorUsuario(int idUsuario)
        {
            return listaPedidos.Where(p => p.getIdUsuario() == idUsuario).ToList();
        }

        /// <summary>
        /// Obtiene la lista de pedidos asociados a un proveedor.
        /// </summary>
        /// <param name="idProveedor">ID del proveedor</param>
        /// <returns>Lista de pedidos del proveedor</returns>
        public List<Pedido> GetListaPedidosPorProveedor(int idProveedor)
        {
            return listaPedidos.Where(p => p.getIdProveedor() == idProveedor).ToList();
        }

        /// <summary>
        /// Obtiene un pedido por su identificador.
        /// </summary>
        /// <param name="idPedido">ID del pedido a buscar</param>
        /// <returns>Pedido encontrado o null si no existe</returns>
        public Pedido GetPedidoPorId(int idPedido)
        {
            return listaPedidos.FirstOrDefault(p => p.getIdPedido() == idPedido);
        }

        /// <summary>
        /// Crea un nuevo pedido.
        /// </summary>
        /// <param name="pedidoNuevo">Pedido a crear</param>
        /// <returns>1 si se creó correctamente, 0 en caso contrario</returns>
        public int CrearPedido(Pedido pedidoNuevo)
        {
            Pedido pedido = new Pedido(
                contadorId++,
                pedidoNuevo.getIdUsuario(),
                pedidoNuevo.getIdProveedor(),
                pedidoNuevo.getFechaPedido(),
                pedidoNuevo.getEstado(),
                pedidoNuevo.getObservaciones(),
                pedidoNuevo.getArchivado()
            );
            listaPedidos.Add(pedido);
            return 1;
        }

        /// <summary>
        /// Actualiza un pedido existente.
        /// </summary>
        /// <param name="idPedido">ID del pedido a actualizar</param>
        /// <param name="pedido">Datos actualizados del pedido</param>
        /// <returns>1 si se actualizó correctamente, 0 si no se encontró</returns>
        public int ActualizarPedido(int idPedido, Pedido pedido)
        {
            Pedido pedidoExistente = listaPedidos.FirstOrDefault(p => p.getIdPedido() == idPedido);
            if (pedidoExistente == null) return 0;

            pedidoExistente.setIdUsuario(pedido.getIdUsuario());
            pedidoExistente.setIdProveedor(pedido.getIdProveedor());
            pedidoExistente.setFechaPedido(pedido.getFechaPedido());
            pedidoExistente.setEstado(pedido.getEstado());
            pedidoExistente.setObservaciones(pedido.getObservaciones());
            return 1;
        }

        /// <summary>
        /// Cambia el estado de un pedido.
        /// </summary>
        /// <param name="idPedido">ID del pedido</param>
        /// <param name="nuevoEstado">Nuevo estado (pedido/enviado/entregado)</param>
        /// <returns>1 si se cambió correctamente, 0 si no se encontró</returns>
        public int CambiarEstadoPedido(int idPedido, string nuevoEstado)
        {
            Pedido pedido = listaPedidos.FirstOrDefault(p => p.getIdPedido() == idPedido);
            if (pedido == null) return 0;

            pedido.setEstado(nuevoEstado);
            return 1;
        }

        /// <summary>
        /// Elimina un pedido por su identificador.
        /// </summary>
        /// <param name="idPedido">ID del pedido a eliminar</param>
        /// <returns>1 si se eliminó correctamente, 0 si no se encontró</returns>
        public int EliminarPedido(int idPedido)
        {
            Pedido pedido = listaPedidos.FirstOrDefault(p => p.getIdPedido() == idPedido);
            if (pedido == null) return 0;

            listaPedidos.Remove(pedido);
            return 1;
        }
    }
}

File: Proyecto-Gruposinco/src/data/repositories/PedidoRepository.ts
import { Injectable } from '@angular/core';
import { clsPedido } from '../../domain/entities/clsPedido';
import { IPedidoRepository } from '../../domain/interfaces/repositories/IPedidoRepository';
import { ApiConnection } from '../datasource/api/ApiConnection';

@Injectable({
  providedIn: 'root'
})
export class PedidoRepository implements IPedidoRepository {

  constructor(private api: ApiConnection) {}

  // Mapper
  private mapToEntity(data: any): clsPedido {
    return new clsPedido(
      data.idPedido,
      data.idUsuario,
      data.idProveedor,
      new Date(data.fechaPedido),
      data.estado,
      data.observaciones ?? "",
      data.archivado ?? false
    );
  }

  // GET listado pedidos
  async getListaPedidos(): Promise<clsPedido[]> {
    const response = await this.api.getPedidos<any[]>();

    if (!response.success || !response.data) {
      throw new Error(response.message);
    }

    return response.data.map(d => this.mapToEntity(d));
  }

  // GET pedidos por usuario
  async getListaPedidosPorUsuario(idUsuario: number): Promise<clsPedido[]> {
    const response = await this.api.getPedidosPorUsuario<any[]>(idUsuario);

    if (!response.success || !response.data) {
      throw new Error(response.message);
    }

    return response.data.map(d => this.mapToEntity(d));
  }

  // GET por proveedor
  async getListaPedidosPorProveedor(idProveedor: number): Promise<clsPedido[]> {
    const pedidos = await this.getListaPedidos();
    return pedidos.filter(p => p.IdProveedor === idProveedor);
  }

  // GET por id
  async getPedidoPorId(idPedido: number): Promise<clsPedido> {
    const response = await this.api.getPedidoPorId<any>(idPedido);

    if (!response.success || !response.data) {
      throw new Error(response.message);
    }

    return this.mapToEntity(response.data);
  }

  // CREATE
  async crearPedido(pedido: clsPedido): Promise<number> {
    const response = await this.api.crearPedido<number>({
      idUsuario: pedido.IdUsuario,
      idProveedor: pedido.IdProveedor,
      fechaPedido: pedido.FechaPedido,
      estado: pedido.Estado,
      observaciones: pedido.Observaciones,
      archivado: pedido.Archivado
    });

    return response.success ? 1 : 0;
  }

  // UPDATE
  async actualizarPedido(idPedido: number, pedido: clsPedido): Promise<number> {
    const response = await this.api.actualizarPedido<number>(
      idPedido,
      {
        idUsuario: pedido.IdUsuario,
        idProveedor: pedido.IdProveedor,
        fechaPedido: pedido.FechaPedido,
        estado: pedido.Estado,
        observaciones: pedido.Observaciones,
        archivado: pedido.Archivado
      }
    );

    return response.success ? 1 : 0;
  }

  // UPDATE estado
  async cambiarEstadoPedido(
    idPedido: number,
    nuevoEstado: "pedido" | "en preparación" | "entregado"
  ): Promise<number> {

    const response = await this.api.actualizarPedido<number>(
      idPedido,
      { estado: nuevoEstado }
    );

    return response.success ? 1 : 0;
  }

  // DELETE
  async eliminarPedido(idPedido: number): Promise<number> {
    const response = await this.api.eliminarPedido<number>(idPedido);
    return response.success ? 1 : 0;
  }
}
